src/order/initial_seg.lean

```diff
@@ -331,8 +337,21 @@ def of_is_empty (r : α → α → Prop) [is_empty α] {b : β} (H : ∀ b', ¬
 @[reducible] def pempty_to_punit : @empty_relation pempty ≺i @empty_relation punit :=
 @of_is_empty _ _ empty_relation _ _ punit.star $ λ x, not_false
 
+protected theorem acc [is_trans β s] (f : r ≺i s) (a : α) : acc r a ↔ acc s (f a) :=
+(f : r ≼i s).acc a
+
 end principal_seg
 
+theorem well_founded_of_principal_seg {β : Type*} (s : β → β → Prop) [is_trans β s]
+  (h : ∀ b, well_founded (subrel s {b' : β | s b' b})) :
+  well_founded s :=
+begin
+  refine ⟨λ b, acc.intro _ (λ b' hb', _)⟩,
+  let f := principal_seg.of_element s b,
+  obtain ⟨b', rfl⟩ := f.down.mp ((principal_seg.of_element_top s b).symm ▸ hb' : s b' f.top),
+  exact (f.acc b').mp ((h b).apply b'),
+end
```

Comment:
Can you explain the name of this lemma? I don't see `principal_seg` anywhere in the statement.

Reply:
I'm not sure about the name, but I have added the docstr.

Reply:
Well, it doesn't hurt to add the real `principal_seg` version of this theorem, but I think it's useless.
- For the backward direction, `principal_seg.of_element` shows these `subrel`s are indeed principal segments.
- For the forward direction, `rel_hom_class.well_founded` is already available for use.